### ISSUE TITLE
Compatibility issues with Octopus

### ISSUE THREAD

**Issue by hschoenburg** [SchemaPlus/schema_monkey#4]
Hello,

I'm seriously struggling to get schema_monkey to work with Octopus ( a gem for database sharding and replication)
https://github.com/tchandy/octopus

Both gems patch ActiveRecord. Together they break. Does SchemaMonkey play well with other gems that monkey patch ActiveRecord? Do I need to rewrite octopus according to the schema_monkey convention for registering modules?

Gemfile
```
gem 'schema_monkey_rails'
gem 'schema_plus_views'
gem 'ar-octopus', require: 'octopus'
```

Here is the error message I'm getting just trying to start the server. 

```
Users/myname/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/schema_monkey-2.0.0/lib/schema_monkey/active_record.rb:6: stack level too deep (SystemStackError)
```

Must admit this is not my usual territory as a developer. Would greatly appreciate some assistance here.

Here is the config/initializers/octopus.rb I'm working from
https://gist.github.com/catsby/6923632
I've tried SchemaMonkey.register Octopus in various places to no avail.

**Comment by johanoskarsson** [comment]
Seeing the same with schema_monkey (2.1.0) and ar-octopus (0.8.5).

Some more of the stacktrace (the last two lines repeat for quite a few times).

SystemStackError (stack level too deep):
  its-it (1.1.1) lib/its-it/it.rb:30:in `each'
  its-it (1.1.1) lib/its-it/it.rb:30:in `inject'
  its-it (1.1.1) lib/its-it/it.rb:30:in `block in to_proc'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:31:in `select'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:31:in `descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `block in descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `each'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `flat_map'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `block in descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `each'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `flat_map'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `block in descendants'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `each'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `flat_map'
  schema_monkey (2.1.0) lib/schema_monkey/module.rb:32:in `descendants'
  schema_monkey (2.1.0) lib/schema_monkey/client.rb:57:in `find_modules'
  schema_monkey (2.1.0) lib/schema_monkey/client.rb:25:in `insert_middleware'
  schema_monkey (2.1.0) lib/schema_monkey/client.rb:11:in `insert'
  its-it (1.1.1) lib/its-it/it.rb:31:in `block (2 levels) in to_proc'
  its-it (1.1.1) lib/its-it/it.rb:30:in `each'
  its-it (1.1.1) lib/its-it/it.rb:30:in `inject'
  its-it (1.1.1) lib/its-it/it.rb:30:in `block in to_proc'
  schema_monkey (2.1.0) lib/schema_monkey/monkey.rb:26:in `each'
  schema_monkey (2.1.0) lib/schema_monkey/monkey.rb:26:in `insert'
  schema_monkey (2.1.0) lib/schema_monkey.rb:33:in `insert'
  schema_monkey (2.1.0) lib/schema_monkey/active_record.rb:11:in `initialize'
  ar-octopus (0.8.5) lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
  schema_monkey (2.1.0) lib/schema_monkey/active_record.rb:13:in `initialize'
  ar-octopus (0.8.5) lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
  schema_monkey (2.1.0) lib/schema_monkey/active_record.rb:13:in `initialize'
....SNIP....
 activerecord (4.2.2) lib/active_record/connection_adapters/postgresql_adapter.rb:227:in `initialize'
  activerecord (4.2.2) lib/active_record/connection_adapters/postgresql_adapter.rb:44:in `new'
  activerecord (4.2.2) lib/active_record/connection_adapters/postgresql_adapter.rb:44:in `postgresql_connection'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:438:in `new_connection'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:448:in `checkout_new_connection'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:422:in `acquire_connection'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:349:in `block in checkout'
  /usr/local/Cellar/ruby/2.2.2/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:348:in `checkout'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:263:in `block in connection'
  /usr/local/Cellar/ruby/2.2.2/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
  activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:262:in `connection'
  ar-octopus (0.8.5) lib/octopus/proxy.rb:205:in `safe_connection'
  ar-octopus (0.8.5) lib/octopus/proxy.rb:212:in `select_connection'
  ar-octopus (0.8.5) lib/octopus/proxy.rb:273:in `method_missing'
...

**Comment by ronen** [comment]
@hschoenburg sorry I never responded, somehow i didn't notice your issue!

@johanoskarsson thanks for the extra stack info.

I think I may have got it:  Octopus looks like it uses some `BasicObject`s, and SchemaMonkey's module descendant crawler would choke when it stumbled on `BasicObject`s because they don't respond to `#is_a?`

I'm not 100% certain that will solve your Octopus problem, but it in any case it was a bug.  I've released schema_monkey v2.1.1 with this fixed.  Try again and see if Octopus works with it now?

**Comment by vanhuyz** [comment]
I've tried schema_monkey v2.1.1 with ar-octopus v0.8.5 today and still no luck.

> stack level too deep (SystemStackError)

**Comment by ronen** [comment]
@vanhyuyz bummer.  can you paste the full stack up to where repeats forever?

**Comment by vanhuyz** [comment]
Here is the stack trace. The last two line repeat forever :sweat_smile: 

> stack level too deep (SystemStackError)
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `group_by'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:57:in `find_modules'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:25:in `insert_middleware'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:11:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:31:in `block (2 levels) in to_proc'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `inject'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `block in to_proc'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/monkey.rb:26:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/monkey.rb:26:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey.rb:33:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/active_record.rb:11:in `initialize'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/active_record.rb:13:in `initialize'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
...

**Comment by vanhuyz** [comment]
Here is the stack trace. The last two lines repeat forever :sweat_smile: 

> stack level too deep (SystemStackError)
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `group_by'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:28:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `block in descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `flat_map'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/module.rb:34:in `descendants'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:57:in `find_modules'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:25:in `insert_middleware'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/client.rb:11:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:31:in `block (2 levels) in to_proc'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `inject'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `block in to_proc'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/monkey.rb:26:in `each'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/monkey.rb:26:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey.rb:33:in `insert'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/active_record.rb:11:in `initialize'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/schema_monkey-2.1.1/lib/schema_monkey/active_record.rb:13:in `initialize'
  /vagrant/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/abstract_adapter.rb:32:in `initialize_with_octopus_shard'
...

**Comment by ronen** [comment]
I found a moment to actually create a test case of my own.. and found that the problem IMHO is with octopus:  it still uses the outdated `alias_method_chain` hack rather than the more modern `Module#prepend` & `super` pattern.  

I've submitted PR tchandy/octopus#321; using octopus from my branch, octopus & schema_plus_views don't get the infinite recursion error.

Caveat: I've only verified that the fix allows me to run migrations and spin up a server; I haven't tried actually using any features of octopus & schema_plus_views together.  But I have no reason to believe they wouldn't work.
